File: src/main/java/com/example/probation/validatiion/validator/PasswordValidator.java
package com.example.probation.validatiion.validator;

import com.example.probation.validatiion.annotation.Password;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

public class PasswordValidator implements ConstraintValidator<Password, String> {
    @Override
    public void initialize(com.example.probation.validatiion.annotation.Password constraintAnnotation) {
        ConstraintValidator.super.initialize(constraintAnnotation);
    }

    @Override
    public boolean isValid(String password, ConstraintValidatorContext constraintValidatorContext) {
        if (password == null) {
            return true;
        }

        return password.matches("(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%]).{8,}$");
    }
}

File: src/main/kotlin/com/example/probation/validation/validator/PasswordValidator.kt
package com.example.probation.validation.validator

import com.example.probation.validation.annotation.Password
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext

class PasswordValidator : ConstraintValidator<Password, String> {
    override fun isValid(password: String?, context: ConstraintValidatorContext) =
        password?.matches("""(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%]).{8,}$""".toRegex()) ?: false

    override fun initialize(password: Password) {
        super.initialize(password)
    }
}
